### PR TITLE
[FIX] test_event_full: fix the registration testing tour

### PR DESCRIPTION
**Pull request by std-odoo**
Bug
===
Sometimes, the registration testing tour failed.

The bug can be semi-deterministic if we add a "sleep(1)" in the endpoint
"/event/<event>/track".

Reason
======
The reason for that is the service worker. It will pre-fetch all the
links in the page (see "prefetch-pages"), so for the "Online Reveal"
we will pre-fetch ~100 pages... If the server is slow, it can cause
issues.

If one endpoint takes some time, all other HTTP requests done by the
service worker will be waiting for it.

So, at the end of the testing tour, the service worker will continue to
make HTTP requests (because it makes the request sequentially) and so
some threads will still be created after the tour.

Even if "_wait_remaining_requests" is called to wait those threads, as
the service worker is still running, it will still continue to make HTTP
requests, creating new threads...

Fix
===
The solution to this issue is to kill the service workers of the browser
when we stop the tour before waiting for the end of the "HTTP request
threads".

Task 2381066

### CODE REVIEW

File: odoo/tests/common.py [hunk 1]
```diff
@@ -1153,6 +1153,16 @@ def navigate_to(self, url, wait_stop=False):
...
+        clear_service_workers = """
+        if ('serviceWorker' in navigator) {
+            navigator.serviceWorker.getRegistrations().then(
+                (registrations) => {registrations.map(r => r.unregister())}
+            )
+        }
+        """
+        cl_id = self._websocket_send('Runtime.evaluate', params={'expression': clear_service_workers})
+        self._websocket_wait_id(cl_id)
+
```
Comment (d-fence) [inline odoo/tests/common.py:1165]: Nice, but did you try this DevTools Protocol command ?
`self._websocket_send('ServiceWorker.stopAllWorkers ')`

This avoid to execute js.
See [Documentation](https://chromedevtools.github.io/devtools-protocol/tot/ServiceWorker/)

**Review by d-fence — CHANGES_REQUESTED** [review]
Nice catch and the fix applies to all HttpCase, nice.
But maybe the DevTools Protocol could be used directly ?

File: odoo/tests/common.py (hunk 1 above)
Comment (std-odoo) [reply odoo/tests/common.py:1165]: @d-fence Hi :)
Good catch for stopAllWorkers, but it doesn't seems to work on the runbot
See https://runbot.odoo.com/runbot/build/5162353

And here's a standard build, all the test passed :) 
https://runbot.odoo.com/runbot/build/5162474

Comment (d-fence) [reply odoo/tests/common.py:1165]: Probably need a [`ServiceWorker.enable`](https://chromedevtools.github.io/devtools-protocol/tot/ServiceWorker/#method-enable) to say that we will use it.

See https://github.com/odoo/odoo/blob/9dc9e00879d018cc9b964eccbda1b6caca480e7e/odoo/tests/common.py#L704-L707

Comment (std-odoo) [reply odoo/tests/common.py:1160]: @d-fence Do I still miss something ? https://runbot.odoo.com/runbot/build/5163476
(work in most cases, but still failed sometimes :/ Like is it didn't always stop the service worker properly)

Comment (d-fence) [reply odoo/tests/common.py:1165]: Ok, we should stick with your first fix as it was working as expected.
The feature is marked as `experimental` ... as most of their features.
This protocol is a moving target, I could try to investigate that in another PR.

**Review by d-fence — APPROVED** [review]

**Comment by tde-banana-odoo** [comment]
If @pparidans agrees, I agree :D . Cannot really tell more for this one ^^ .

**Comment by d-fence** [comment]
@std-odoo please just remove the `TO REMOVE` commit so that we can merge it

**Comment by std-odoo** [comment]
@tde-banana-odoo 
@d-fence 
Done :)

**Comment by pparidans** [comment]
@d-fence @std-odoo @tde-banana-odoo Do we all agree that this PR entirely skip the Service Workers in the test, and so they aren't tested anymore, AT ALL!
Meaning that the SW can be broken once deployed and still have a green runbot !

I do agree that the prefetch could create unintentional load/requests during tests (and a way to mitigate that is interresting)... but skipping the SW entirely is dangerous, I think...

Also, @std-odoo, there may be a misunderstanding about the "sequentiallity" of the requests going through the SW : only the trigger of the prefetched url's requests are made sequentially, not the processing of ALL requests in the SW (at least they shouldn't)!

**Comment by tde-banana-odoo** [comment]
Could we skip Service Workers globally in tests but have some dedicated tests for them to ensure they work ? Or are they somehow app-specific ?

**Comment by std-odoo** [comment]
> @d-fence @std-odoo @tde-banana-odoo Do we all agree that this PR entirely skip the Service Workers in the test, and so they aren't tested anymore, AT ALL!
> Meaning that the SW can be broken once deployed and still have a green runbot !
> 
> I do agree that the prefetch could create unintentional load/requests during tests (and a way to mitigate that is interresting)... but skipping the SW entirely is dangerous, I think...
> 
> Also, @std-odoo, there may be a misunderstanding about the "sequentiallity" of the requests going through the SW : only the trigger of the prefetched url's requests are made sequentially, not the processing of ALL requests in the SW (at least they shouldn't)!
> 
> TLDR: I see the point, but this fix looks like a "hide the dust under the carpet" kind of fix

No, it's done in the **clear** method, so when we close the chrome headless. So the service worker is running during the tests and killed when the testing tour is done.

Also, I understood that all the requests are done sequentially, but it's still an issue for me, not for the tests, but for the "real usage".
It means that even if the visitor close the tab, we will continue to fetch 100 pages... (because it's done in the service worker) so basically, all visitor will download ~the entire event website after the first loading...
BTW it's not the purpose of the task, but I just wanted to say what I think about this...

**Comment by std-odoo** [comment]
(To resume, SW is running during the tests and killed after the tour is done)

**Comment by d-fence** [comment]
@pparidans the clear method is called at the end of the browser_js method ... after the tour is finished.
So I think that does not impact the tour, or the tour is not well crafted and needs to continue ?
The only drawback I see, is that when the browser_js method is called again in the same test, the service worker are now stopped.
So maybe we could ensure that this service is started at the beginning ?

**Comment by std-odoo** [comment]
In that case, the service worker will restart from the beginning.
But we also clear the cookies, the local storage, etc so I think it's OK to restart the SW.
What do you think ?

**Comment by d-fence** [comment]
> In that case, the service worker will restart from the beginning.
> But we also clear the cookies, the local storage, etc so I think it's OK to restart the SW.
> What do you think ?

Do you mean stop it and immediately restart it in the clear method ?

**Comment by std-odoo** [comment]
@d-fence No, when we will reload the page, we will fetch the SW code, and because no SW is running, it will start (like if we were loading the page for the first time).
But if we really want to keep the SW running between all the `browse_js` call in the tests, we should not clear the local storage.

And I don't really understand why it's an issue to restart the SW between the `browse_js` call... I means it's generally 2 different tours no ?

**Comment by std-odoo** [comment]
(might be run by 2 different users, etc)

**Comment by pparidans** [comment]
@d-fence @std-odoo Thanks for the clarification about the "clear()" method.

I still think that the SW should be working during tests... but could (should?) be restarted between each run (actually if the Chrome profile is dropped between runs, the SW shouldn't be alive, no?).

**Comment by pparidans** [comment]
> (might be run by 2 different users, etc)

@d-fence I guess each run of Chrome Headless is done on its own profile ?

**Comment by d-fence** [comment]
I just read some doc about the service worker. Sorry I didn't know how it worked.
I seems that it starts on demand, so the fix looks ok no ?
You unregister the service worker for this tour (like we clear the storage, cookies...) and if another tour is run by the browser_js method in the same test, the service worker will be restarted by the tour if needed.

**Comment by std-odoo** [comment]
@pparidans the [SWs are working during the tests](https://github.com/odoo/odoo/pull/62248#issuecomment-737146237), and indeed, each run (`browser_js` call) has its "own profile" (own cookies, local storage, SW...)

**Comment by pparidans** [comment]
> @pparidans the [SWs are working during the tests](https://github.com/odoo/odoo/pull/62248#issuecomment-737146237), and indeed, each run (`browser_js` call) has its "own profile" (own cookies, local storage, SW...)

I agree with that. And you PR looks ok in that regard (I needed the clarification about `clear()`).

What I'm wondering is if (and why) the SW is not killed when Chrome is closed and its profile destroyed (which should give the same result: no more requests triggered)?

File: odoo/tests/common.py [hunk 2]
```diff
@@ -1153,6 +1153,16 @@ def navigate_to(self, url, wait_stop=False):
             self._websocket_wait_event('Page.frameStoppedLoading', params={'frameId': frame_id})
 
     def clear(self):
+        clear_service_workers = """
+        if ('serviceWorker' in navigator) {
+            navigator.serviceWorker.getRegistrations().then(
+                (registrations) => {registrations.map(r => r.unregister())}
```
Comment (pparidans) [inline odoo/tests/common.py:1159]: Why use a `map()` as you just need to iterate over registrations, not transform them?

**Comment by d-fence** [comment]
> > (might be run by 2 different users, etc)
> 
> @d-fence I guess each run of Chrome Headless is done on its own profile ?

A small clarification here:
Each ChromeBrowser use a [separate datadir](https://github.com/odoo/odoo/blob/53f877bf38caa2c8fafcd35392c5e0e278ce5937/odoo/tests/common.py#L808) meaning a separate profile.
But a the same running browser can be re-used in the same HttpCase test by calling the browser_js method again.
That's why we `reset` the browser state in with the clear method.

So, most of the time, the tests are run with a really separated profile.

File: odoo/tests/common.py [hunk 3]
```diff
@@ -1153,6 +1153,16 @@ def navigate_to(self, url, wait_stop=False):
...
+            )
+        }
+        """
+        cl_id = self._websocket_send('Runtime.evaluate', params={'expression': clear_service_workers})
+        self._websocket_wait_id(cl_id)
```
Comment (pparidans) [inline odoo/tests/common.py:1164]: `clear_service_workers` snippet use a Promise (from `getRegistrations()`) but doesn't wait for it (at least in JS).

Is it normal? Is there something on the Python side that waits ?

Technically speaking, the snippet will finish its execution immediately even if the `unregister()` call(s) are not completed. In our case, that shouldn't long but still...

**Comment by d-fence** [comment]
> What I'm wondering is if (and why) the SW is not killed when Chrome is closed and its profile destroyed (which should give the same result: no more requests triggered)?

As I explained, the browser is kept running until the end of the python test case. It may be re-used in another test method.
In [the test we try to fix](https://github.com/odoo/odoo/blob/82b06dc4c3df502c8e006169f4ee0f6aa33e3ac1/addons/test_event_full/tests/test_wevent_register.py#L13) some python assertions are made before stopping the browser. It seems that it's enough to have the service worker starting new requests in the meanwhile.

File: odoo/tests/common.py (hunk 3 above)
Comment (pparidans) [reply odoo/tests/common.py:1163]: Maybe a detail but you can use a `const` instead of a `let` as this variable isn't reassigned.

Comment (pparidans) [reply odoo/tests/common.py:1163]: Also, about the `await`, I guess that as this snippet is executed in the global context, so it works even outside an `async` ?

Comment (std-odoo) [reply odoo/tests/common.py]: @pparidans Indeed for const :) About await it worked well, but let's try the option `awaitPromise`

Comment (pparidans) [reply odoo/tests/common.py]: Actually, found the release notes I was thinking about :wink: 
https://developers.google.com/web/updates/2017/08/devtools-release-notes#await

Top level await is supported only in the DevTools (which we use in this case), so it's fine using it here.

Comment (pparidans) [reply odoo/tests/common.py]: Not sure about the `awaitPromise` as your expression doesn't return anything...

Comment (std-odoo) [reply odoo/tests/common.py]: Yes, you might be right, thanks

Comment (std-odoo) [reply odoo/tests/common.py:1164]: @pparidans Mhé, awaiting like this cause sometimes an issue -_-
See https://runbot.odoo.com/runbot/build/5249178

So I think I'll comeback to my original fix which worked all the times...

Comment (pparidans) [reply odoo/tests/common.py:1164]: Is it really the "good fix" to not wait the promise... :thinking:

Comment (std-odoo) [reply odoo/tests/common.py]: @pparidans It seems that waiting for it do not always fix the problem... But I absolutely don't know why...
Keeping the old code and adding the option "awaitPromise" also work if you really want to wait the promise.

Comment (pparidans) [reply odoo/tests/common.py]: I really want to wait for it :wink: 

LGTM :+1:

**Review by pparidans — APPROVED** [review]

**Comment by d-fence** [comment]
then it's time for
robodoo r+

**Comment by pparidans** [comment]
@std-odoo @d-fence Force-pushed after the r+ ?!? :warning:

**Comment by pparidans** [comment]
Why moving that code down the `clear()` method ?

**Comment by d-fence** [comment]
??

**Comment by tde-banana-odoo** [comment]
:lobster:

**Comment by std-odoo** [comment]
It feel cleaner / better in that place, sorry, I didn't see that we already r+

**Comment by pparidans** [comment]
> It feel cleaner / better in that place, sorry, I didn't see that we already r+

Well, by doing it after the cleanup of cookies, storage... you change the context of the SW *before* killing it!

I think it could cause unpredictable results if the SW still attempts to perform requests (e.g. requests triggered before and after the cleanup won't run in the same "environment").

**Comment by std-odoo** [comment]
@pparidans Well, I moved it before clearing cookie / cache ;)

**Comment by pparidans** [comment]
Looks ok to me!
If the Runbot agrees, let's merge this :+1:

**Comment by d-fence** [comment]
robodoo r+
Thanks guys
